### PR TITLE
More cleanups in Assembly/Loader/Binder area

### PR DESCRIPTION
**Pull request by VSadov**
- [x]  fold the native `AssemblyLoadContext` into its base `AssemblyBinder`
Now `DefaultAssemblyBinder` and `CustomAssemblyBinder` derive directly from `AssemblyBinder`
- [x]  rename: NativeAssemblyLoadContext -> NativeAssemblyBinder
- [x]  misc. renames: 
TPABinderContext -> DefaultBinder,   FallbackLoadContextBinder -> FallbackBinder, etc ...
- [x]  remove `CoreBindResult` and everything under `src/coreclr/vm/coreclr`  folder
The type is basically a tuple of `BINDER_SPACE::Assembly` and `HRESULT` and `HRESULT` is used in a single place right after it is set. The rest is COM support, ref-counting, holder, init/reset/delete, and other plumbing. We can just use `Assembly` directly.

### CODE REVIEW

**Review by vitek-karas — APPROVED** [review]
Looks good.

File: src/coreclr/vm/appdomain.cpp [hunk 1]
```diff
@@ -3695,15 +3695,12 @@ PEAssembly * AppDomain::BindAssemblySpec(
...
-                // Use CoreClr's fusion alternative
-                CoreBindResult bindResult;
+                ReleaseHolder<BINDER_SPACE::Assembly> bindResult;
+                hrBindResult = pSpec->Bind(this, &bindResult);
 
-                pSpec->Bind(this, FALSE /* fThrowOnFileNotFound */, &bindResult);
-                hrBindResult = bindResult.GetHRBindResult();
-
-                if (bindResult.Found())
+                if (bindResult)
```
Comment (vitek-karas) [inline src/coreclr/vm/appdomain.cpp:3701]: Nit: I would rename the local variable to something like `boundAssembly` - basically include the word "Assembly" in it. It feels weird to read the code below and use a "result" as an assembly. (Not that it was better before, it was even worse)

**Review by vitek-karas — APPROVED** [review]

File: src/coreclr/vm/appdomain.cpp (hunk 1 above)
Comment (VSadov) [reply src/coreclr/vm/appdomain.cpp:3701]: Right. I actually thought that `boundAssembly` could be a better name here. `pBinder` would be another choice. I am not a fan of 'p' prefixes, but it is used in many places in the binder code. 
Anyways, `boundAssembly`  feels better here.

I think I will make another pass over variables typed as `AssemblyBinder*`. We have a variety of [ load | context | binder ] combinations when naming these. Perhaps calling them `binder` or `pBinder`, in majority of cases at least, would make it more consistent.

**Review by janvorli — COMMENTED** [review]
The change looks good modulo the bug.

File: src/coreclr/vm/assembly.cpp [hunk 2]
```diff
@@ -543,14 +543,14 @@ Assembly *Assembly::CreateDynamic(AppDomain *pDomain, AssemblyBinder* pBinderCon
             pAssemblyLoaderAllocator->SetupManagedTracking(&args->loaderAllocator);
             createdNewAssemblyLoaderAllocator = TRUE;
 
-            if(pBinderAssemblyLoaderAllocator != nullptr)
+            if(pAssemblyLoaderAllocator != nullptr)
             {
-                pAssemblyLoaderAllocator->EnsureReference(pBinderAssemblyLoaderAllocator);
+                pAssemblyLoaderAllocator->EnsureReference(pAssemblyLoaderAllocator);
```
[the remark this answers is not in the document]
Comment (VSadov) [reply src/coreclr/vm/assembly.cpp:548]: oh, right, thee are two of them here. Thanks!

File: src/coreclr/vm/assemblynative.cpp [hunk 3]
```diff
@@ -54,12 +54,12 @@ void QCALLTYPE AssemblyNative::InternalLoad(QCall::ObjectHandleOnStack assemblyN
...
     DomainAssembly * pParentAssembly = NULL;
     Assembly * pRefAssembly = NULL;
-    AssemblyBinder *pBinderContext = NULL;
+    AssemblyBinder *pBinder = NULL;
 
     if (assemblyLoadContext.Get() != NULL)
     {
-        INT_PTR nativeAssemblyLoadContext = ((ASSEMBLYLOADCONTEXTREF)assemblyLoadContext.Get())->GetNativeAssemblyLoadContext();
-        pBinderContext = reinterpret_cast<AssemblyBinder*>(nativeAssemblyLoadContext);
+        INT_PTR nativeAssemblyLoadContext = ((ASSEMBLYLOADCONTEXTREF)assemblyLoadContext.Get())->GetNativeAssemblyBinder();
```
[the remark this answers is not in the document]
Comment (VSadov) [reply src/coreclr/vm/assemblynative.cpp:61]: Yes, nativeAssemblyBinder will be better.

**Comment by VSadov** [comment]
Thanks!!!

**Pull request merged**